clang-tools-extra/docs/clang-tidy/checks/google/cpp-init-class-members.rst

```diff
@@ -0,0 +1,32 @@
+.. title:: clang-tidy - google-cpp-init-class-members
+
+google-cpp-init-class-members
+=============================
+
+Checks that class members are initialized in constructors (implicitly or
+explicitly). Reports constructors or classes where class members are not
+initialized. The goal of this checker is to eliminate UUM (Use of
+Uninitialized Memory) bugs caused by uninitialized class members.
+
+This checker is different from ProTypeMemberInitCheck in that this checker
+attempts to eliminate UUMs as a bug class, at the expense of false
+positives.
```

Comment:
It's not entirely clear to me what "UUMs as a bug class" means. Could you provide an example where `ProTypeMemberInitCheck` doesn't detect the problem, and this check does? Also, why shouldn't that be fixed in `ProTypeMemberInitCheck` instead?

Reply:
# Examples 
>  Could you provide an example where ProTypeMemberInitCheck doesn't detect the problem, and this check does? 

The 2 cases for which this PR is written (available in the unit tests) are not found by `ProTypeMemberInitCheck`.

Specifically (I copy them from the unit test):

```
class PositiveDefaultedDefaultConstructor {
public:
  PositiveDefaultedDefaultConstructor() = default;
  // CHECK-MESSAGES: :[[@LINE-1]]:3: warning: constructor should initialize these fields: X

private:
  int X;
};
```

and 

```
struct PositiveStruct {
  // CHECK-MESSAGES: :[[@LINE-1]]:8: warning: these fields should be initialized: X, Y
  int X;
  int Y;
};
```

Also a variation (again, copied them from the unit test)

```
class PositiveDefaultedConstructorObjectAndPrimitive {
 public:
  PositiveDefaultedConstructorObjectAndPrimitive() = default;
  // CHECK-MESSAGES: :[[@LINE-1]]:3: warning: constructor should initialize these fields: Y

  Helper* GetHelper() { return &X; }

  void SetY(bool enabled) { Y = enabled; }

  bool IsY() { return Y; }

 private:
  Helper X;
  bool Y;
};

# Regarding adding to the existing checker.

I fully discuss this in the [comment below](https://github.com/llvm/llvm-project/pull/65189#issuecomment-1709185690). Please take a look there for the full context. Thank you!

Reply:
First example is fine, but for that exist check should be just extended.
Second example is not so fine, simply because this is POD, and by adding initialization you change this, and this may impact usage of such class, like zeroed using memset, used in union. For me if you want to force initialization for such classes, this should be separated check or option in exist one, maybe distinguish based on keyword struct vs class.
Last issue is already detected by cppcoreguidelines-pro-type-member-init.